packages/jellyfish-api-core/__tests__/category/account/futureswap.test.ts

```diff
@@ -0,0 +1,1093 @@
+      await expect(promise).rejects.toThrow(RpcApiError)
+      await expect(promise).rejects.toThrow('RpcApiError: \': Invalid amount\', code: -3, method: futureswap')
+    }
+    {
+      // invlaid source dtoken
+      const fswap: FutureSwap = {
+        address: tslaAddress,
+        amount: '1@INVALID'
+      }
+      const promise = testing.rpc.account.futureSwap(fswap)
+      await expect(promise).rejects.toThrow(RpcApiError)
+      await expect(promise).rejects.toThrow('RpcApiError: \': Invalid Defi token: INVALID\', code: 0, method: futureswap')
+    }
+    {
+      // non loan source token
+      const fswap: FutureSwap = {
+        address: tslaAddress,
+        amount: '1@BTC'
+      }
+      const promise = testing.rpc.account.futureSwap(fswap)
+      await expect(promise).rejects.toThrow(RpcApiError)
+      await expect(promise).rejects.toThrow('RpcApiError: \'Test DFIP2203Tx execution failed:\nCould not get source loan token 1\', code: -32600, method: futureswap')
+    }
+    {
+      // destination is given when futureswap dtoken to dusd
+      const fswap: FutureSwap = {
+        address: tslaAddress,
+        amount: '1@TSLA',
+        destination: 'DUSD'
+      }
+      const promise = testing.rpc.account.futureSwap(fswap)
+      await expect(promise).rejects.toThrow(RpcApiError)
+      await expect(promise).rejects.toThrow('RpcApiError: \'Test DFIP2203Tx execution failed:\nDestination should not be set when source amount is a dToken\', code: -32600, method: futureswap')
+    }
+    {
+      // INVALID destination is given when futureswap dusd to dtoken
+      const fswap: FutureSwap = {
+        address: tslaAddress,
+        amount: '1@DUSD',
+        destination: 'INVALID'
+      }
+      const promise = testing.rpc.account.futureSwap(fswap)
+      await expect(promise).rejects.toThrow(RpcApiError)
+      await expect(promise).rejects.toThrow('RpcApiError: \'Destination token not found\', code: -5, method: futureswap')
+    }
+    {
+      // arbitary address without enough balance
+      const fswap: FutureSwap = {
+        address: await testing.generateAddress(),
+        amount: '1@TSLA'
+      }
+      const promise = testing.rpc.account.futureSwap(fswap)
+      await expect(promise).rejects.toThrow(RpcApiError)
+      await expect(promise).rejects.toThrow('RpcApiError: \'Test DFIP2203Tx execution failed:\namount 0.00000000 is less than 1.00000000\', code: -32600, method: futureswap')
+    }
+  })
+
+  it('should not create futureswap when DFIP2203 is disabled for the dtoken', async () => {
+    const swapAmount = 1
+    const tslaAddress = await testing.generateAddress()
+    await testing.rpc.account.accountToAccount(collateralAddress, { [tslaAddress]: `${swapAmount}@TSLA` })
+    await testing.generate(1)
+
+    // deactivate DFIP2203 for dtoken
+    const key = `v0/token/${idTSLA}/dfip2203`
+
+    await testing.rpc.masternode.setGov({ [attributeKey]: { [key]: 'false' } })
```

Comment:
Here, If we set false for DUSD instead of TSLA, below `futureSwap` call will proceed without exception, is it acceptable ? (Same thing happened for `withdrawFutureSwap`)

Reply:
cpp code only checks for the dtoken. either dtoken -> dusd or dusd -> dtoken. Ref - https://github.com/DeFiCh/ain/blob/e63ae12e399b846c5f97dbf3cb1e771f75263255/src/masternodes/mn_checks.cpp#L1500
Thinking about the business use cases, i guess it's stil fine since you can only do swap with a dtoken. and checking the dtoken is enough.
@prasannavl